### ISSUE TITLE
SMARTY ForceField topology representation

### ISSUE THREAD

**Issue by jchodera** [open-forcefield-group/smarty#18]
This is a question for @cbayly13, @davidlmobley, @pgrinaway, and others to chime in on.

My original goal was to replace the OpenMM `simtk.openmm.app.ForceField` class with a works-alike class, but I am wondering if we want to have it accept a different kind of input.

Currently, we use `ForceField` in the following way:

``` python
from simtk.openmm.app import PDBFile, ForceField
# Load a PDB file
pdbfile = PDBFile('mypdb.pdb')
# Create a ForceField object for AMBER 99SB-ILDN and TIP3P
forcefield = ForceField('amber99sbildn.xml', 'tip3p.xml')
# Create an OpenMM System object from the topology in the PDB file
# Other options (e.g. nonbonded method) are specified in ...
system = forcefield.createSystem(pdbfile.topology, ...)
```

The `Topology` stores atom names, residue names, connectivity, etc., but the only information that `createSystem` uses is (1) the elements associated with each atom, and (2) the connectivity (without bond order).

For the SMARTY version of `ForceField` we will use the SMIRKS matching facility.  I'm actually not entirely sure exactly how to use the SMIRKS matching in the same way we might use the SMILES matching (any insight @cbayly13?), but if we were matching SMARTS, we would use a piece of code like this:

``` python
# create a SMARTS search object
ss = OEMCSSearch(smarts)
OEPrepareSearch(mol, ss)
# loop over matches
for count, match in enumerate(ss.Match(mol)):
   # do stuff here...
```

where `mol` would be an `OEMol` representation of the system. 

Things I'm unsure of:
- We likely can't easily represent _many_ molecules in a single `OEMol`, so we'd need to enumerate each molecule in `Topology` and parameterize them separately. For systems that contain many copies of the same molecule, this would be efficient, so we'd want to cache the parameters for each unique molecule type.
- If we just have elements and connectivity (without bond order), it is actually incredibly difficult to create an `OEMol` object. I've been cheating and using the AMBER `bondtype` in [this code](https://github.com/choderalab/openmoltools/blob/master/openmoltools/forcefield_generators.py#L80-L179), but I have a feeling it is fragile and suboptimal. Should we extend `Topology` to incorporate bond order? Or should we use an entirely different representation than an OpenMM `Topology` to start with that might be better suited to systems containing small molecules and biopolymers?

**Comment by davidlmobley** [comment]
Let me start by asking for some clarification on a couple things just to make sure I don't miss any important points:

> The Topology stores atom names, residue names, connectivity, etc., but the only information that createSystem uses is (1) the elements associated with each atom, and (2) the connectivity (without bond order).

So are you saying we'd use `createSystem` here as well, but we would write a new `createSystem` which would use SMIRKS matching instead of how it operates presently? And, if needed, would our new `createSystem` be able to use more information from the `Topology` if needed? 

At this point you're loading from a PDB file:
`pdbfile = PDBFile('mypdb.pdb')`

Could this be modified to load an alternate format, i.e. with bond orders? I believe we were planning a canonical format of `.sdf` with custom tags. Though, this format may not make sense for representing an entire system at once, which may be why you're talking about starting from a `.pdb` here...? 

*_This point ^ is going to get more involved and touch your second question, so I'll get to it below in a separate comment. *_

> We likely can't easily represent many molecules in a single OEMol, so we'd need to enumerate each molecule in Topology and parameterize them separately. For systems that contain many copies of the same molecule, this would be efficient, so we'd want to cache the parameters for each unique molecule type

Just to clarify - you mean **inefficient** above, not **efficient**, right? 

I would expect ParmEd can perhaps help here; it has a convenient splitting functionality (i.e. see https://github.com/ParmEd/ParmEd/issues/450) where one can take a system object and split it into components; it recognizes repeated residues. So you could simply load your topology into ParmEd, split it to recognize individual repeated units (it will tell you how many counts of each there are), assign parameters once for each, then re-merge and move into OpenMM... (?)

**Comment by davidlmobley** [comment]
OK, to get to your final point:

> If we just have elements and connectivity (without bond order), it is actually incredibly difficult to create an OEMol object. I've been cheating and using the AMBER bondtype in this code, but I have a feeling it is fragile and suboptimal. Should we extend Topology to incorporate bond order? Or should we use an entirely different representation than an OpenMM Topology to start with that might be better suited to systems containing small molecules and biopolymers?

Indeed! This gets back to the issue of whether a `PDB` can really be a suitable starting point, and my answer is **NO!**. We **really really\* need to not just start with elements and connectivity, as this is terrible. My sense (and I'm guessing Christopher's as well (I'll talk to him about it in the morning and report back), and certainly that of many people with lots of pharma experience such as Pat Walters, who I asked directly about this) is that you, sadly, currently need **two files** or two formats to specify the contents of a general biomolecular system you want to simulate. One is a format which says exactly what each thing IS, and the other is a format which says exactly where each thing is and how many there are. Currently the best thing we seem to have is a `PDB` file that says where everything is/what the full contents of the system is, and an `.sdf` file (really, one for each component) with custom tags (such as for bond order, etc.) that specifies what each component is. You can handle some substitutions - i.e., the `PDB` can be replaced with a variety common molecular simulation formats, and the `.sdf` could (for small molecules) often be replaced by SMILES, `.mol2`, etc.

OK, but back to your specific questions... 

> Should we extend Topology to incorporate bond order? Or should we use an entirely different representation than an OpenMM Topology to start with that might be better suited to systems containing small molecules and biopolymers?

I think my answer is that we have to do at least one of these, and I'll discuss this with the group on this end in the morning. My first response is to say we ought to do something that is deliberately tailored to systems containing small molecules, arbitrary mixtures of solvents, and biopolymers, but _maybe_ it is easy enough to extend `Topology` to get what we want? 

I'm just worried about what will happen if we get stuck in a world where we are forced to load things from a single file format which is sub-optimal (such as `.pdb`) rather than two file formats (.sdf and .pdb or some such) which really seems to be what we need... 

I'll report back after I discuss with the group.

**Comment by jchodera** [comment]
> So are you saying we'd use `createSystem` here as well, but we would write a new `createSystem` which would use SMIRKS matching instead of how it operates presently? 

Currently, OpenMM provides a very-convenient `ForceField` class that operates in the way I described: It provides a `createSystem(topology, ...)` method that creates a new `System` object from a `Topology` object.  It currently _only_ uses the atomic elements and connectivity (without bond order) information.

My original goal was to provide a drop-in replacement for `ForceField` that is initialized with the SMARTY XML format. This would provide the maximum flexibility to use it anywhere you would normally use OpenMM, but would constrain us to _only_ have the elements and connectivity available for defining the molecular system, which can cause problems taking molecules into `OEMol`.

> And, if needed, would our new createSystem be able to use more information from the Topology if needed?

If needed, yes, but there isn't much more information in there that would be of use, since atom and residue names and numbers are arbitrary.  We've discussed [adding optional bond orders](https://github.com/pandegroup/openmm/issues/1443), which is the simplest extension that would solve the problem of unambiguously creating `OEMol`s from molecules in the file, but @cbayly13 seems to want to move beyond bond orders (though how he would change all of the OpenEye toolkit to function without them, I am not certain).

> At this point you're loading from a PDB file:
> Could this be modified to load an alternate format, i.e. with bond orders? 

Yes, absolutely.  The important function signature is the `createSystem(topology, ...)`, where the `Topology` object is sufficient.

> I believe we were planning a canonical format of .sdf with custom tags. Though, this format may not make sense for representing an entire system at once, which may be why you're talking about starting from a .pdb here...?

Would you use a multi-molecule SDF file? For example, would there be a different molecule in the file for each molecule in the simulation?

> Just to clarify - you mean inefficient above, not efficient, right?

Correct!

> I would expect ParmEd can perhaps help here; it has a convenient splitting functionality (i.e. see ParmEd/ParmEd#450) where one can take a system object and split it into components; it recognizes repeated residues. So you could simply load your topology into ParmEd, split it to recognize individual repeated units (it will tell you how many counts of each there are), assign parameters once for each, then re-merge and move into OpenMM... (?)

I'll look into this.

> Currently the best thing we seem to have is a PDB file that says where everything is/what the full contents of the system is, and an .sdf file (really, one for each component) with custom tags (such as for bond order, etc.) that specifies what each component is. You can handle some substitutions - i.e., the PDB can be replaced with a variety common molecular simulation formats, and the .sdf could (for small molecules) often be replaced by SMILES, .mol2, etc.

We could, for now, require that `createSystem` take _both_ an OpenMM `Topology` and a list of `OEMol` objects that match the connectivity and elements of residues in `Topology`.  The `OEMol`s would be used for typing and the `Topology` would be used for figuring out what order they appear in the system. 

``` python
from simtk.openmm.app import PDBFile
from smarty import ForceField
# Load a PDB file to specify where molecules appear in the system
pdbfile = PDBFile('mypdb.pdb')
# Load OEMol molecules to define bond order and connectivity (in arbitrary order)
# TODO: Figure out if this is the best idiom for loading multiple OEGraphMols into a set
mols = list()
ifs = oemolistream()
mol = OEGraphMol()
while OEReadMolecule(ifs, mol):
   mols.append( OEGraphMol(mol) )
# Create a ForceField object for AMBER 99SB-ILDN and TIP3P
forcefield = ForceField('amber99sbildn.xml', 'tip3p.xml')
# Create an OpenMM System object from the topology in the PDB file
# Other options (e.g. nonbonded method) are specified in ...
system = forcefield.createSystem(pdbfile.topology, molecules=mols, ...)
```

This would work fine for systems of only small molecules and solvents, but I don't know if this would work for biopolymers unless you can really load a whole protein as an `OEMol` from an SDF.  This is where @cbayly13's insight about biopolymer handling would help.

**Comment by jchodera** [comment]
One other _critical question for @cbayly13 that I forgot to highlight_:

For the SMARTY version of ForceField we will use the SMIRKS matching facility. I'm actually not entirely sure exactly how to use the SMIRKS matching in the same way we might use the SMILES matching.

_Is there an equivalent of this idiom (for SMARTS) that works for SMIRKS too, and how do we determine which atoms match which numbers in the SMIRKS string?_

``` python
# create a SMARTS search object
ss = OEMCSSearch(smarts)
OEPrepareSearch(mol, ss)
# loop over matches
for count, match in enumerate(ss.Match(mol)):
   # do stuff here...
```

I basically need to know _how I iterate over SMIRKS matches_ using the OEChem toolkit.

**Comment by davidlmobley** [comment]
I'll go over all of this with Christopher as soon as we are both in today, likely around 9 am Pacific but I am not totally sure when he gets in. Are you available for a call in the early afternoon your time (noon, 1, or 1:30) if it's helpful to discuss briefly? 

On this one point:

> We could, for now, require that createSystem take both an OpenMM Topology and a list of OEMol objects that match the connectivity and elements of residues in Topology. The OEMols would be used for typing and the Topology would be used for figuring out what order they appear in the system.

This would be ideal from a small molecule/solutions perspective, certainly. I don't **think** there is any fundamental problem with using this for biomolecules as well (the OEDocking toolkit represents the protein as an OEMol or OEGraphMol typically), so for example from some docking code I have which I adapted from an OpenEye example:

``` python
    receptor = OEGraphMol()
    imstr = oemolistream('3NF8_prepped.pdb')
    protein = OEGraphMol()
    OEReadMolecule(imstr, protein)
    imstr.close()
    imstr = oemolistream('3NF8_ligand.pdb')
    ligand = OEGraphMol()
    OEReadMolecule(imstr, ligand)
    OEMakeReceptor(receptor, protein, ligand)
```

**Comment by davidlmobley** [comment]
i.e. see https://docs.eyesopen.com/toolkits/python/dockingtk/receptor.html#creating-a-receptor in the docs

**Comment by davidlmobley** [comment]
FTR, since I'd forgotten - it seems the main difference between `OEMol` and `OEGraphMol` that I can find offhand is that the former is multi-conformer.

**Comment by jchodera** [comment]
I should be available anytime after 1.30P EST for ad-hoc video calls or slack chats.

**Comment by davidlmobley** [comment]
We'll be in touch, @jchodera .

**Comment by davidlmobley** [comment]
We're discussing, @jchodera , and I can get you outcomes shortly or we can talk briefly by phone and I can document here afterwards. 

@cbayly13 says that OEParseSmarts understands SMIRKS; apparently there was an internal debate at OE about whether these deserved a separate name and the "no they don't" group won. 

However, he has never used this exact type of functionality:

``` python
# create a SMARTS search object
ss = OEMCSSearch(smarts)
OEPrepareSearch(mol, ss)
# loop over matches
for count, match in enumerate(ss.Match(mol)):
   # do stuff here...
```

so he can't comment on whether it will work here. He's trying to dig up an example of what he HAS done.

**Comment by jchodera** [comment]
Any examples would be appreciated!

Feel free to just summarize the discussion unless you need to discuss with me.

**Comment by davidlmobley** [comment]
@jchodera - OK, he has found his example - which is a minor variant of his `smartsDepict.ipynb` which illustrates how to do the matching you want and we're going to post on GitHub and reference here as soon as we have it up. (PS it is now listed above).

I'll also answer the outcome to this:

> We could, for now, require that createSystem take both an OpenMM Topology and a list of OEMol objects that match the connectivity and elements of residues in Topology. The OEMols would be used for typing and the Topology would be used for figuring out what order they appear in the system.

Our answer at this point is that we want `createSystem` to take an OpenMM topology and a set of `.sdf` files with custom tags (including fractional bond order, partial charge and so on) for the components. We're in agreement on this end that you need the topology which gives you coordinates, etc., for the full system, and that will be critical for large biomolecules like proteins, but we also really need the exact chemistry for the individual pieces. I can talk to you briefly as to why `OEMols` are not the best choice for this when you have a minute. 

I realize the above statement is saying that `createSystem` will be reading files, which may not be what you want. If that's the case let us know and we can revisit whether it  might be OK for it to take `OEMol` objects for the components after all. But that's our first answer.

**Comment by jchodera** [comment]
> Our answer at this point is that we want createSystem to take an OpenMM topology and a set of .sdf files with custom tags (including fractional bond order, partial charge and so on) for the components. 

I don't think we should have it take _files_, but rather a set of `OEGraphMol`s with the appropriate associated tags.  

We needed a documented specification for these tags, @davidlmobley.

>  I can talk to you briefly as to why OEMols are not the best choice for this when you have a minute.

I'm confused about this. Do you expect me to write my own SDF reader? I'm definitely _not_ going to do that. We should invent a different format instead that suits our needs if we're not going to go through `OEMol`.

> I realize the above statement is saying that createSystem will be reading files, which may not be what you want. 

That is correct. We don't want this.

> If that's the case let us know and we can revisit whether it might be OK for it to take OEMol objects for the components after all. But that's our first answer.

Under the hood, `ForceField` will need to create `OEGraphMol` objects for each of the molecules in the topology, so we will _need_ to get the molecules into `OEGraphMol` format eventually. Requiring SDF files be passed limits our flexibility.

**Comment by davidlmobley** [comment]
We discussed this by phone (the three of us) and decided that we want to extend the `OpenMM` `Topology` to include bond order, which will get us the information we need in order to be able to proceed without having to rely on `SDF` files. 

Internally within `createSystem`, though, we will have to take the `Topology` object into `OEMol` objects for components, so we will need to create a tool which does this. I will create a separate issue for this unless, @jchodera , you already have something which would do this?

Also, @cbayly13 remarks that creating a molecule from connectivity, element type, and bond order in OE world usually also requires **formal charges** so he thinks **the `Topology` object needs to include formal charges as well** at least for now. We're going to investigate with OpenEye to see if the formal charge can be perceived from the other information in the case where all atoms are explicit (which is what we have here) but he doesn't know if there is a tool which does this.

**Comment by davidlmobley** [comment]
@jchodera - is there a way this tool can be done before the "Topology-to-OEMol" issue is resolved, or do you already have a solution to that issue that we're unaware of? This is the main hurdle remaining it seems to us.

Though, since this will be done internally within `createSystem`, perhaps a hack-ish interim solution is permissible if we plan to swap in something better later - for example, presumably an `OpenMM` topology can be split using `ParmEd`, the components written out to (for example) `.mol2` or `.pdb` files, and then these turned into OEMol objects. For initial tests this may be sufficient.

**Comment by jchodera** [comment]
I have an interim solution for small molecules, I think. Running a bit
behind, but should be able to finish this week.

**Comment by davidlmobley** [comment]
I think this is closed for now by #24
